base/methodshow.jl

```diff
@@ -301,7 +301,8 @@ function show_method_table(io::IO, ms::MethodList, max::Int=-1, header::Bool=tru
     last_shown_line_infos === nothing || empty!(last_shown_line_infos)
 
     modul = if mt === _TYPE_NAME.mt && length(ms) > 0 # type constructor
-            which(ms.ms[1].module, ms.ms[1].name)
+            i = findfirst(m -> isdefined(m.module, m.name), ms.ms)
+            i === nothing ? mt.module : which(ms.ms[i].module, ms.ms[i].name)
```

Comment:
```suggestion
            sig = Base.unwrap_unionall(ms[1].sig)::DataType # Tuple{Type{...}}
            ((sig.parameters[1]::DataType).parameters[1]::DataType).name.module
```
This seems very hacky, but would be more correct?

Reply:
I would think most of those type asserts are invalid, because the general assumption here (that method tables are owned by modules) is invalid itself

Reply:
Looking over the code a bit closer, it looks like it wants to check it the method itself is being defined on a function / type in the same module or in a dependency. But that needs to be a per-method query, and is not valid to derive from the method table

Reply:
I guess the difficulty is defining what "same module" means here since you need to select a starting point for comparison in order to cycle the colors. @aviatesk's suggestion does seem better than what I've implemented, except it uses `ms[1]` as on current master, the defining module of which isn't necessarily meaningful. But it at least shouldn't error since it isn't using `which` as on master.

Reply:
Combining part of @aviatesk's suggestion with part of what's currently implemented in this PR:
```suggestion
            i = something(findfirst(m -> isdefined(m.module, m.name), ms.ms), 1)
            parentmodule(fieldtype(unwrap_unionall(ms[i].sig), 1).parameters[1])
```
Does this seem reasonable, @vtjnash?

Reply:
It is a bit better. The more generally correct version of this needs to be done for each method however, first attempting to derive the method table it is being added to, then deriving if it is the same as the method itself. Something like this:
```suggestion
mmt = get_methodtable(meth)
modulecolor = if mmt.module === parentmodule(meth)
        nothing
    else
        # mmt is only particularly relevant for external method tables. Since the primary method table is shared, we now need to distinguish "primary" methods by trying to check if there is a primary DataType to identify it with
        # c.f. how `jl_method_def` would derive this same information (for the name)
        ft = argument_datatype((unwrap_unionall(meth.sig)::DataType).parameters[1]) # try to get the type associated with the first argument
        isType(ft) && (ft = argument_datatype(ft.parameters[1])) # if the ft is `Type`, try to unwrap it again
        if ft === nothing || parentmodule(meth) === parentmodule(ft) !== Core
            nothing
        else
            # get color for meth
        end
end
...
```

n.b. In C, the first `ft` computation is called  `jl_nth_argument_datatype`, but it seems we don't export that to Julia